# Dominic's Spotify API Report (15 October 2023)

I plan to take one of the playlists from my Spotify library and analyze the danceability, energy, and liveliness of the playlist and understand the differences in the type of music I listen to. I think it would be intriguing to compare this data across the music and songs contained in this playlist. 

Theoretical Hypothesis - I have a playlist (The Ultimate Mixtape) on Spotify that contains 478 songs (28 hrs and 12 mins). I hypothesize that the danceability, energy, and liveliness will be on the higher end as many of the songs on the playlist include more types of genres like pop, country, and rock.

Statistical Hypothesis - I hypothesize that the danceability, energy, and liveliness will contain higher numbers as compared to that of acoustics and instrumentals.  

The source and structure of the data used in this report originated from Spotify for Developers. This is a useful tool presented directly from Spotify that allows the user to utilize Web API to retrieve information from the Spotify catalog about artists, tracks, and playlists. It contains large amounts of data that are useful for working with.

In [1]:
import requests
import pandas as pd
import base64
import json
import urllib

In [2]:
Client_ID = pd.read_csv('Spotify_IDs')['Client_ID'].iloc[0]

In [3]:
Client_Secret = pd.read_csv('Spotify_IDs')['Client_Secret'].iloc[0]

In [4]:
# Client_ID

In [5]:
# Client_Secret

In [6]:
client_cred = base64.b64encode(str(Client_ID + ":" + Client_Secret).encode("ascii"))

In [7]:
# client_cred

In [8]:
headers = {"Authorization": "Basic {}".format(client_cred.decode("ascii"))}

In [9]:
# headers

In [10]:
payload = {'grant_type' : 'client_credentials'}
url = 'https://accounts.spotify.com/api/token'

In [11]:
session_key_response = requests.post(url = url, data = payload, headers = headers)

In [12]:
session_key_response.status_code

200

In [13]:
# session_key_response.json()['access_token']

In [14]:
session_header_key = session_key_response.json()

In [15]:
key = session_header_key['access_token']

In [16]:
session_headers = {"Authorization": "Bearer {}".format(key)}

In [17]:
# session_headers

In [18]:
search_url = 'https://api.spotify.com/v1/playlists/54iGOfVAwrFokblOdcUAjB'

In [19]:
search_url

'https://api.spotify.com/v1/playlists/54iGOfVAwrFokblOdcUAjB'

In [20]:
search_response = requests.get(url = search_url, headers = session_headers)
search_response.status_code

200

In [21]:
search_data = search_response.json()

In [22]:
search_data.keys()

dict_keys(['collaborative', 'description', 'external_urls', 'followers', 'href', 'id', 'images', 'name', 'owner', 'primary_color', 'public', 'snapshot_id', 'tracks', 'type', 'uri'])

In [23]:
search_data['tracks'].keys()

dict_keys(['href', 'items', 'limit', 'next', 'offset', 'previous', 'total'])

In [24]:
search_data['tracks']['items'][0].keys()

dict_keys(['added_at', 'added_by', 'is_local', 'primary_color', 'track', 'video_thumbnail'])

In [25]:
search_df = pd.DataFrame(search_data['tracks']['items'])

In [26]:
search_df.head()

added_at                                           added_by  \
0  2023-08-20T20:16:26Z  {'external_urls': {'spotify': 'https://open.sp...   
1  2023-08-20T20:16:26Z  {'external_urls': {'spotify': 'https://open.sp...   
2  2023-08-20T20:16:26Z  {'external_urls': {'spotify': 'https://open.sp...   
3  2023-08-20T20:16:26Z  {'external_urls': {'spotify': 'https://open.sp...   
4  2023-08-20T20:16:26Z  {'external_urls': {'spotify': 'https://open.sp...   

   is_local primary_color                                              track  \
0     False          None  {'album': {'album_type': 'album', 'artists': [...   
1     False          None  {'album': {'album_type': 'album', 'artists': [...   
2     False          None  {'album': {'album_type': 'album', 'artists': [...   
3     False          None  {'album': {'album_type': 'album', 'artists': [...   
4     False          None  {'album': {'album_type': 'album', 'artists': [...   

  video_thumbnail  
0   {'url': None}  
1   {'url': None}  
2   {'url': None}  
3   {'url': None}  
4   {'url': None}

In [27]:
search_df['track'].iloc[0]['id']

'2SiXAy7TuUkycRVbbWDEpo'

In [28]:
ids = [x['id'] for x in search_df['track']]

In [29]:
','.join(ids)

'2SiXAy7TuUkycRVbbWDEpo,57bgtoPSgt236HzfBOd8kj,0rmGAIH9LNJewFw7nKzZnc,2zYzyRzz6pRmhPzyfMEC8s,7ACxUo21jtTHzy7ZEV56vU,0G21yYKMZoHa30cYVi1iA8,1hlveB9M6ijHZRbzZ2teyh,4JfuiOWlWCkjP6OKurHjSn,58XWGx7KNNkKneHdprcprX,7N3PAbqfTjSEU1edb2tY8j,6kooDsorCpWVMGc994XjWN,6gQUbFwwdYXlKdmqRoWKJe,5KaR60ue2myx5Y1ImL0STA,18ESXa5mEm1V4Pkt5GSXWx,6qUEOWqOzu1rLPUPQ1ECpx,0rFWuqFgHAfuzE8uSB9TWR,5u5F7qLDvZjBSktaDp4HxB,549Go7a66CX6k523uK9kUv,1gzIbdFnGJ226LTl0Cn2SX,4KfSdst7rW39C0sfhArdrz,00qOE7OjRl0BpYiCiweZB2,73TxYZd0lBCVRrHawrAglA,2HXixVqzzm9rEUIMAWzshl,35k31HZI4z9PbBOioaI4dZ,0vOkmmJEtjuFZDzrQSFzEE,6xMHglHoafdDFGXS6qfwSH,5B6Z7NwVqA7W3UFiqBdEJC,2oSpQ7QtIKTNFfA08Cy0ku,2ccUQnjjNWT0rsNnsBpsCA,1ynmMEK1fkyiZ6Z6F3ThEt,7vidktgNZFQylTgH1GEnMs,3Pu3IChaAoa5EjgUBv5F6x,3pHST9YijKAzBJlIRS7R47,2cOpRVjsE4qZLZbFOB47VJ,0powdVQEIZSaSnNupt2yPE,4XIdxJBiWXUaQiJY5hlSIx,47ji0KfmvI93CMDqppVj8p,3n5TN5s5NFWatzrsFmQOse,2jockdEsw2U8Vf1HWBXWIc,3RsL0gYDlwaZWxHJJwO3j8,2dMjy70ufAcxK2ib7pEW4n,76iyBw2Q1HAnnR2SVYa6iL,4kpewfJPc1FWCuOMaDha1k,5ilbsx6qkd

In [30]:
audio_features_url = 'https://api.spotify.com/v1/audio-features?ids={}'.format(','.join(ids))

In [31]:
audio_features_url

'https://api.spotify.com/v1/audio-features?ids=2SiXAy7TuUkycRVbbWDEpo,57bgtoPSgt236HzfBOd8kj,0rmGAIH9LNJewFw7nKzZnc,2zYzyRzz6pRmhPzyfMEC8s,7ACxUo21jtTHzy7ZEV56vU,0G21yYKMZoHa30cYVi1iA8,1hlveB9M6ijHZRbzZ2teyh,4JfuiOWlWCkjP6OKurHjSn,58XWGx7KNNkKneHdprcprX,7N3PAbqfTjSEU1edb2tY8j,6kooDsorCpWVMGc994XjWN,6gQUbFwwdYXlKdmqRoWKJe,5KaR60ue2myx5Y1ImL0STA,18ESXa5mEm1V4Pkt5GSXWx,6qUEOWqOzu1rLPUPQ1ECpx,0rFWuqFgHAfuzE8uSB9TWR,5u5F7qLDvZjBSktaDp4HxB,549Go7a66CX6k523uK9kUv,1gzIbdFnGJ226LTl0Cn2SX,4KfSdst7rW39C0sfhArdrz,00qOE7OjRl0BpYiCiweZB2,73TxYZd0lBCVRrHawrAglA,2HXixVqzzm9rEUIMAWzshl,35k31HZI4z9PbBOioaI4dZ,0vOkmmJEtjuFZDzrQSFzEE,6xMHglHoafdDFGXS6qfwSH,5B6Z7NwVqA7W3UFiqBdEJC,2oSpQ7QtIKTNFfA08Cy0ku,2ccUQnjjNWT0rsNnsBpsCA,1ynmMEK1fkyiZ6Z6F3ThEt,7vidktgNZFQylTgH1GEnMs,3Pu3IChaAoa5EjgUBv5F6x,3pHST9YijKAzBJlIRS7R47,2cOpRVjsE4qZLZbFOB47VJ,0powdVQEIZSaSnNupt2yPE,4XIdxJBiWXUaQiJY5hlSIx,47ji0KfmvI93CMDqppVj8p,3n5TN5s5NFWatzrsFmQOse,2jockdEsw2U8Vf1HWBXWIc,3RsL0gYDlwaZWxHJJwO3j8,2dMjy70ufAcxK2ib7pEW4n,76iyBw2Q1H

In [32]:
features_response = requests.get(url = audio_features_url, headers = session_headers)
features_response.status_code

200

In [33]:
features = features_response.json()

In [34]:
features_df = pd.DataFrame(features['audio_features'])

In [35]:
features_df.head(50)

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.532   0.767    7    -5.509     1       0.0574      0.002870   
1          0.502   0.890    4    -5.175     1       0.0364      0.000147   
2          0.541   0.960    0    -2.947     0       0.0505      0.066300   
3          0.574   0.913    6    -4.793     0       0.1330      0.061000   
4          0.454   0.913    9    -4.008     1       0.1070      0.408000   
5          0.453   0.987    8    -4.485     1       0.0889      0.019500   
6          0.502   0.924    4    -6.299     1       0.0503      0.012700   
7          0.653   0.730    5   -10.613     1       0.0429      0.011500   
8          0.482   0.617    4   -12.931     0       0.0458      0.005430   
9          0.572   0.835    0    -6.219     1       0.0317      0.171000   
10         0.465   0.907   11    -6.404     0       0.1130      0.024100   
11         0.540   0.982    4    -2.962     0       0.0458      0.435000   
12         0.448   0.908   11    -3.818     0       0.1470      0.032100   
13         0.407   0.955    2    -4.164     1       0.1640      0.025100   
14         0.761   0.711    9    -7.989     1       0.0604      0.017400   
15         0.262   0.777   11    -8.671     1       0.0632      0.010900   
16         0.649   0.903    7    -3.677     1       0.0318      0.000841   
17         0.599   0.931    7    -4.653     1       0.0735      0.030200   
18         0.673   0.814    9    -5.411     1       0.0367      0.013300   
19         0.548   0.691    4   -13.553     0       0.0369      0.015000   
20         0.357   0.653    9    -5.554     1       0.0654      0.082800   
21         0.379   0.760    9   -10.961     1       0.0378      0.002980   
22         0.611   0.882    9   -11.225     1       0.0538      0.084700   
23         0.515   0.766    7   -12.768     1       0.0343      0.024900   
24         0.741   0.580    4    -9.050     1       0.0304      0.117000   
25         0.705   0.806    0    -4.830     1       0.0814      0.229000   
26         0.628   0.934   11    -6.026     1       0.0527      0.006820   
27         0.381   0.894    7    -8.638     1       0.0370      0.025800   
28         0.524   0.955    0    -7.081     1       0.0411      0.007170   
29         0.653   0.677    0   -12.244     1       0.0396      0.249000   
30         0.562   0.609    3    -7.063     0       0.0271      0.026400   
31         0.644   0.799    4   -10.856     0       0.0598      0.082100   
32         0.351   0.889    8    -3.513     1       0.1150      0.001620   
33         0.616   0.910    8    -3.316     0       0.1240      0.010900   
34         0.593   0.814    9    -4.802     1       0.1660      0.019700   
35         0.664   0.785    8    -4.985     1       0.0688      0.016600   
36         0.588   0.860    8    -4.461     1       0.0767      0.028000   
37         0.614   0.840    9    -4.624     1       0.0410      0.006890   
38         0.614   0.792    0    -4.476     0       0.0636      0.040700   
39         0.791   0.796   10    -6.630     1       0.0901      0.021000   
40         0.557   0.796    3    -4.147     1       0.0482      0.004110   
41         0.531   0.881    9    -3.909     1       0.0402      0.003180   
42         0.586   0.896    3    -4.572     1       0.1100      0.114000   
43         0.613   0.917    8    -4.754     1       0.0637      0.000398   
44         0.623   0.869    5    -3.376     1       0.0481      0.008410   
45         0.591   0.866    2    -2.877     1       0.0536      0.057600   
46         0.541   0.910    8    -4.725     1       0.0520      0.000927   
47         0.568   0.865    5    -3.653     1       0.0486      0.006320   
48         0.664   0.739    5    -5.342     0       0.0713      0.006840   
49         0.581   0.860   10    -3.459     1       0.0360      0.000501   

    instrumentalness  liveness  valence    tempo            type  \
0           0.000513    0.3900    0.755  127.361  audio_fea

After achieving the data from this playlist and comparing each song with its danceability, energy, and liveliness versus its acoustics and instrumentals, I can make conclusions and comparisons about my hypothesis. 

First, after working with the data for some time, I concluded that my hypothesis had the potential to be stronger. While the hypothesis wasn't "incomplete" I believe I could have developed a stronger hypothesis instead of using phrases such as, "on the higher end" and "contain higher numbers." While my hypothesis helps make conclusions and comparisons, I had the potential to make it stronger and develop a deeper and more robust hypothesis which may have allowed me to go beyond the surface level of the data. Nevertheless, the data I was able to produce allows me to answer the hypothesis as it was stated. 

Secondly, I am not afraid to admit that my professor helped and supported me through the gathering and development of this data. Based on his examples and lessons, I was able to work through and process this data. As I am working with the Spotify data, there are important pieces of information that are hidden to safeguard my data and information. However, with this information, I can work through a process of taking my playlist (The Ultimate Mixtape) and parsing out important keys to break down the internal data and, by using list comprehension, make inferences about the data. While there are limitations to the data, some resulting from my moderate hypothesis and others from my lack of experience working with this type of data, the information that I do have allows me to understand, on some level, the information on this playlist and inspire me to devel deeper into doing more with the data. Also, by being able to make charts and tables, I could simplify the data I would like to compare even more to make it more concise and organized. 

Nevertheless, the data I gained from this playlist allowed me to take the songs listed in my playlist, and compare the danceability, energy, and liveliness versus its acoustics and instrumentals within the song itself and with other songs listed. For example, song number 0 had a danceability of 0.532, energy of 0.767, and liveliness of 0.3900 as compared to its acoustics of 0.002870 and instrumentalness of 0.000513. If I were to compare this song with itself, I can determine that it supports my hypothesis that danceability, energy, and liveliness contain "higher numbers" as compared to that of acoustics and instrumentalness. 

I can also support my hypothesis that danceability, energy, and liveliness contain "higher numbers" as compared to that of acoustics and instrumentalness between the songs and not just one song. In other words, if I were to compare these statistics in song 1 and song 2:

Song 1 contains 0.502(danceability), 0.890(energy), and 0.2170(liveliness) versus its 0.000147(acoustics) and 0.011700(instrumentalness). 

Song 2 contains 0.541(danceability), 0.960(energy), and 0.3440(liveliness), versus its 0.066300(acoustics) and	0.000002(instrumentalness).

By comparing these two songs I can determine that they have "higher numbers." in danceability, energy, and liveliness. I believe if I were to cross reference all songs listed in the data frame, I could take the data, and by determining the name of the songs listed, I can support the theory that the country, rock, and pop songs, are the songs resulting in higher numbers. 

While there may be limitations to this data, the information that I have developed allows me to understand the data and make inferences and conclusions.